Start guidance on tshirt sizing (#1680)
grommet/hpe-design-system: aries-site/src/pages/foundation/tshirt-sizing.mdx

import { Box, Image } from 'grommet';
import { Example } from '../../layouts';
import { ButtonSizingExample } from '../../examples';

## What is t-shirt sizing?

T-shirt sizing is the way that Grommet leverages familiar clothing sizes of "small", "medium", "large", and other ranges of sizes to drive the sizing and spacing of components. Grommet maps these t-shirt sizes to specific pixel values. The derivation of these pixel values will be explained more below.

> T-shirt sizing controls two aspects of a UI: the width/height of components and the spacing within and around components (margin, pad, border, etc.)

Reviewer: We also use t-shirt concepts in font-sizing and responsive breakpoints. Debating whether they should be mentioned here or does that clutter the messaging? If mentioned, they could link to Typography and ResponsiveContext content.

When designing and developing with the HPE Design System, you should always use t-shirt sizes. If you find yourself adding custom px values to any properties, this should be a warning that you need to consult with your designer about why a custom value is needed. In nearly every case, designs and developement should be using t-shirt sizes.

## Why do we use t-shirt sizing?

Using t-shirt sizes results in a UI that is more consistent across pages and across applications, creating a more unified experience for HPE customers.

There are also benefits of t-shirt sizing for designers and developers!

> Using pixel values, percentages, or other common sizing methods to create layouts can result in applications that are less composable and less scalable. T-shirt sizing creates a common language between designers and developers and defined intervals of sizes that makes designing and developing more efficient. Instead, a container of content may just have of a width of "medium", for example, and Grommet will take care of the rest.

Reviewer: ```suggestion
T-shirt sizing creates a common language between designers and developers, as well as provides common, well-defined intervals which make designing and developing more efficient.
Using pixel values, percentages, or other sizing methods to create layouts can result in applications that are less composable and less scalable. Instead, a container of content may just have a width of "medium", for example, and Grommet will take care of the rest.

Finally, t-shirt sizes enable content to responsively scale more easily and consistently. For example, a layout's padding may be set to "large" on desktop screen and gracefully adjust to "small" on mobile devices with minimal effort.
```

## The base unit

In order to create the range of sizes and their corresponding pixel values, a base unit of `24px` is used as a scalar.

> Below is a snippet from Grommet's [`base.js`](https://github.com/grommet/grommet/blob/master/src/js/themes/base.js) theme file where the calculations and corresponding pixel values for the t-shirt sizes that drive the width/height of components such as Box are defined. The `base.js` theme establishes many of the defaults that are leveraged across Grommet applications.

Reviewer: ```suggestion
Below is a snippet from Grommet's [`base.js`](https://github.com/grommet/grommet/blob/master/src/js/themes/base.js) theme file where the calculations and resulting pixel values for the t-shirt sizes driving the height/width for components such as [Box](/components/box) are defined. The `base.js` theme establishes many of the defaults leveraged across Grommet applications.
```

```
// base.js in Grommet

> 

Reviewer: Should `const baseSpacing = 24;` be defined in this snippet for someone who is visually skimming?
```suggestion
const baseSpacing = 24;

...
```

> xxsmall: `${baseSpacing * 2}px`, // 48
> xsmall: `${baseSpacing * 4}px`, // 96
> small: `${baseSpacing * 8}px`, // 192
> medium: `${baseSpacing * 16}px`, // 384
> large: `${baseSpacing * 32}px`, // 768
> xlarge: `${baseSpacing * 48}px`, // 1152
> xxlarge: `${baseSpacing * 64}px`, // 1536

Reviewer: ```suggestion
xxsmall: `${baseSpacing * 2}px`, // 48px
xsmall: `${baseSpacing * 4}px`, // 96px
small: `${baseSpacing * 8}px`, // 192px
medium: `${baseSpacing * 16}px`, // 384px
large: `${baseSpacing * 32}px`, // 768px
xlarge: `${baseSpacing * 48}px`, // 1152px
xxlarge: `${baseSpacing * 64}px`, // 1536px
```

full: '100%',
```

## T-shirt sizing for component dimensions

> One of the main ways t-shirt sizing is used is to control the width/height of a component. For some components, like [Box](/components/box), this can be controlled with the `width` and `height` props. For other components, like [Avatar](https://v2.grommet.io/avatar?theme=hpe), [Table](/components/table), or [Text](/foundation/typography#text), the dimensions are controlled by the `size` prop.

Reviewer: ```suggestion
One of the main ways t-shirt sizing is used is to control the width/height of a component. For some components, like [Box](/components/box), this can be controlled with the `width` and `height` props. For other components, like [Avatar](https://v2.grommet.io/avatar?theme=hpe), [Table](/components/table), or [Text](/foundation/typography#text), the dimensions are controlled by the `size` prop. And others, like [Grid](/components/grid), t-shirt sizes specify `row` and `colum` dimensions.
```

> The [Grommet site](https://v2.grommet.io/) will provide documentation on the props available to any given component. If you're unsure if a component supports `width`, `height`, or `size`, that would be the right place to check.

Reviewer: ```suggestion
The [Grommet site](https://v2.grommet.io/) provides documentation on the props available to any given component. If you're unsure if a component supports `width`, `height`, or `size`, that would be the right place to check.
```

Given the code snippet from the previous section and the explanation on t-shirt sizing for dimensions, if you were to implement the following code, you would have a Box with width of "384px" and height of "192px".

```
<Box width="medium" height="small" />
// width = 384px, height = 192px
```

<Box width="large">
  <Image src="/box-sizing.png" alt="Demonstration of t-shirt sizing on Box" />
</Box>

## T-shirt sizing for spacing and other styles

In addition to overall component dimensions, t-shirt sizing can also be used to define spacing and other styling such as pad, gap, margin, and border.

To create the pixel values for these styles, Grommet divides and/or multiplies the base unit of `24px` by various integers.

Pad, gap, and margin are controlled by `edgeSize` and the corresponding pixel values also come from [`base.js`](https://github.com/grommet/grommet/blob/master/src/js/themes/base.js) in Grommet

```
// base.js in Grommet

> 

Reviewer: Define `const baseSpacing = 24;`?

```suggestion

const baseSpacing = 24;

....

```

edgeSize: {
    none: '0px',
    hair: '1px', // for Chart
    xxsmall: `${baseSpacing / 8}px`, // 3
    xsmall: `${baseSpacing / 4}px`, // 6
    small: `${baseSpacing / 2}px`, // 12
    medium: `${baseSpacing}px`, // 24
    large: `${baseSpacing * 2}px`, // 48
    xlarge: `${baseSpacing * 4}px`, // 96

}
```

Given this code snippet, if you were to implement the following code, you would have a Box with width of 384px, height of 192px, and pad of 48px.

```
<Box
    width="medium"
    height="small"
    pad="large"
/>
```

<Box width="medium" margin={{bottom: 'large'}}>
  <Image
    src="/edge-sizing.png"
    alt="Demonstration of t-shirt sizing for pad on Box"
  />
</Box>

Border sizing is controlled by `borderSize`.

```
// base.js in Grommet

> 

Reviewer: Define `const baseSpacing = 24;`?

```suggestion

const baseSpacing = 24;

....

```
